clippy_lints/src/splitting_strings_at_newlines.rs

```diff
@@ -0,0 +1,89 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for usages of `str.trim().split("\n")` and `str.trim().split("\r\n")`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Hard-coding the line endings makes the code less portable. `str.lines` should be used instead.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".trim().split('\n');
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".lines();
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub SPLITTING_STRINGS_AT_NEWLINES,
+    suspicious,
+    "splitting a trimmed string at hard-coded newlines"
+}
+
+declare_lint_pass!(SplittingStringsAtNewlines => [SPLITTING_STRINGS_AT_NEWLINES]);
+
+// Check if an expr is of a type that implements `Deref` with `Target = str`
+fn derefs_into_str(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
+    matches!(
+        cx.tcx.get_diagnostic_item(sym::Deref),
+        Some(def_id) if (
+            implements_trait(cx, cx.typeck_results().expr_ty(expr), def_id, &[])
+            && matches!(
+                cx.get_associated_type(cx.typeck_results().expr_ty_adjusted(expr), def_id, "Target"),
+                Some(target) if target.is_str()
+            )
+        )
+    )
+}
+
+impl LateLintPass<'_> for SplittingStringsAtNewlines {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        // We're looking for `A.trim().split(B)`, where the type of `A` implements `Deref` with `Target =
+        // str` (e.g. a `str` literal or an expression returning `String`), and `B` is a `Pattern`
+        // that hard-codes a newline (either `"\n"` or `"\r\n"`). There are a lot of ways to specify a
+        // pattern, and this lint only checks the most basic ones: a `'\n'`, `"\n"`, and `"\r\n"`.
+
+        if expr.span.from_expansion() {
```

Comment:
Maybe this lint could live in the `methods` subdir? It does these kinds of checks for you

Reply:
I'm not sure what the exact conventions are for that. The Clippy book [mentions those "types"](https://doc.rust-lang.org/nightly/clippy/development/adding_lints.html#specific-type) but doesn't really say when to choose them. 

Taking a closer look, there's already a `methods.str_splitn`, so maybe it would make sense to move the new lint there or (since it's not strictly about `splitn`) to something like `methods.str_split`?

But on the other hand, `strlen_on_c_strings` is just in `clippy_lints`, and not in a subdir.

I don't have any strong preference, so feel free to suggest what you think is the best place.

Reply:
AFAIK, this isn't really "enforced" at this point (and there's no real observable difference to users in which lint pass a lint is defined, I think), but there have been some efforts (#11493) to move lints to more general lint passes, since apparently it can have a performance impact, but is also just nicer since a bunch of checks can be shared between all method lints.

I think having it be a new file in `methods/` sounds like a good idea. Different lints are usually in different files unless they are very tightly coupled/depend on each other (eg look for very similar patterns and need additional logic to avoid double-linting the same expression)

Reply:
I moved the lint to a new `methods/str_strip.rs` file, modeled after the existing `methods/str_splitn.rs`. I also adjusted the name of the lint itself slightly to hopefully better match the existing ones -- it is now `STR_SPLIT_AT_NEWLINE`.